### PR TITLE
Feature/1071/get-PromoCode-By-Code

### PR DESCRIPTION
**Pull request by LubomyrKashuba**
## Description

Added posibility to query promocode by code name

### Checklist
- [x] 🔽 My branch is up-to-date with "development" branch
- [x] ✅All tests passed locally
- [x] ✨My changes working with up-to-date front-end and admin part locally, like charm
- [x] 🔗 Link pull request to issue

### CODE REVIEW

File: modules/promo-code/promo-code.permissions.js [hunk 1]
```diff
@@ -1,9 +1,14 @@
...
+const promoCodeQuery = {
+  getAllPromoCodes: allow,
+  getPromoCodeByCode: allow,
+};
+
 const promoCodeMutation = {
   addPromoCode: allow,
   deletePromoCode: allow,
   updatePromoCode: allow,
 };
```
Comment (AndriyLymych) [inline modules/promo-code/promo-code.permissions.js:12]: why have all these methods been allowed?
not logged persons can see all promocodes?
please, describe requirements

File: modules/promo-code/promo-code.resolver.js [hunk 2]
```diff
@@ -4,6 +4,9 @@ const promoCodeQuery = {
   getAllPromoCodes: () => PromoCodeService.getAllPromoCodes(),
 
   getPromoCodeById: async ({ id }) => PromoCodeService.getPromoCodeById(id),
+
+  getPromoCodeByCode: async (_, { code }) =>
+    PromoCodeService.getPromoCodeByCode(code),
```
Comment (AndriyLymych) [inline modules/promo-code/promo-code.resolver.js:9]: why `PromoCodeService`? it isn't class name, just an instance. just `promoCodeService`

File: tests/promo-code/promo-code.query.test.js [hunk 3]
```diff
@@ -8,4 +13,11 @@ describe('Promo-codes queries', () => {
     expect(answers).toBeDefined();
     expect(answers.items).toBeInstanceOf(Array);
   });
+
+  it('Should get promo-code by field code', async () => {
+    const operations = await setupApp();
+    await addPromoCode(newPromoCode, operations);
+    const promoCode = await getPromoCodeByCode(newPromoCode.code, operations);
```
Comment (AndriyLymych) [inline tests/promo-code/promo-code.query.test.js:20]: empty line

File: typeDefs.js [hunk 4]
```diff
@@ -560,6 +560,7 @@ const typeDefs = gql`
       filter: MaterialFilterInput,
     ): PaginatedMaterials!
     getPromoCodeById(id: ID): PromoCode
+    getPromoCodeByCode(code: String!): PromoCode
```
Comment (AndriyLymych) [inline typeDefs.js:563]: please attach here how PromoCode fields looks

File: tests/promo-code/promo-code.query.test.js (hunk 3 above)
Comment (AndriyLymych) [reply tests/promo-code/promo-code.query.test.js:20]: please, test also case when promocode is not found

File: typeDefs.js (hunk 4 above)
Comment (AndriyLymych) [reply typeDefs.js:563]: if in all promocodes methods error case not handled, need to create task for refactoring

File: modules/promo-code/promo-code.permissions.js (hunk 1 above)
Comment (LubomyrKashuba) [reply modules/promo-code/promo-code.permissions.js:16]: I got it, in Progress

File: modules/promo-code/promo-code.resolver.js (hunk 2 above)
Comment (LubomyrKashuba) [reply modules/promo-code/promo-code.resolver.js]: will be done

File: tests/promo-code/promo-code.query.test.js (hunk 3 above)
Comment (LubomyrKashuba) [reply tests/promo-code/promo-code.query.test.js:27]: Ok, got it

File: typeDefs.js (hunk 4 above)
Comment (LubomyrKashuba) [reply typeDefs.js]: In progress

File: modules/promo-code/promo-code.permissions.js (hunk 1 above)
Comment (LubomyrKashuba) [reply modules/promo-code/promo-code.permissions.js:16]: Done

File: modules/promo-code/promo-code.resolver.js (hunk 2 above)
Comment (LubomyrKashuba) [reply modules/promo-code/promo-code.resolver.js]: Fixed

File: tests/promo-code/promo-code.query.test.js (hunk 3 above)
Comment (LubomyrKashuba) [reply tests/promo-code/promo-code.query.test.js:27]: Done

File: typeDefs.js (hunk 4 above)
Comment (LubomyrKashuba) [reply typeDefs.js]: Done

File: tests/promo-code/promo-code.query.test.js [hunk 5]
```diff
@@ -1,11 +1,39 @@
...
+
+    expect(promoCode).toHaveProperty('code', newPromoCode.code);
+  });
+
+  it('Should receive Error', async () => {
+    const operations = await setupApp();
+    const promoCode = await getPromoCodeByCode('error', operations);
+
+    expect(promoCode).toHaveProperty('message', PROMOCODE_NOT_FOUND);
+    expect(promoCode).toHaveProperty('statusCode', NOT_FOUND);
```
Comment (AndriyLymych) [inline tests/promo-code/promo-code.query.test.js:37]: you dont need 2 expects here

just one common

File: modules/promo-code/promo-code.resolver.js [hunk 6]
```diff
@@ -1,19 +1,22 @@
...
   addPromoCode: async (_, { promoCode }) =>
-    PromoCodeService.addPromoCode(promoCode),
+    promoCodeService.addPromoCode(promoCode),
 
-  deletePromoCode: async (_, args) => PromoCodeService.deletePromoCode(args.id),
+  deletePromoCode: async (_, args) => promoCodeService.deletePromoCode(args.id),
 
   updatePromoCode: async (_, args) =>
-    PromoCodeService.updatePromoCode(args.id, args.promoCode),
+    promoCodeService.updatePromoCode(args.id, args.promoCode),
```
Comment (AndriyLymych) [inline modules/promo-code/promo-code.resolver.js:19]: where do you catch your errors?

File: modules/promo-code/promo-code.service.js [hunk 7]
```diff
@@ -1,5 +1,13 @@
...
+const {
+  STATUS_CODES: { NOT_FOUND },
+} = require('../../consts/status-codes');
+const {
+  PROMOCODE_NOT_FOUND,
+} = require('../../error-messages/promocode.messages');
+
 class PromoCodeService {
   async getAllPromoCodes() {
     const items = await PromoCode.find().exec();
```
Comment (AndriyLymych) [inline modules/promo-code/promo-code.service.js:13]: and...that's all? pagination for this?

Comment (AndriyLymych) [reply modules/promo-code/promo-code.service.js:13]: or create task for refactoring

File: modules/promo-code/promo-code.permissions.js (hunk 1 above)
Comment (AndriyLymych) [reply modules/promo-code/promo-code.permissions.js:16]: please, give me requirements for these methods

**Review by maarekua — APPROVED** [review]

**Review by tarasSanotskyi — APPROVED** [review]

**Pull request merged**